src/dependency_injector/_cwiring.pyx

```diff
@@ -62,25 +62,24 @@ async def _async_inject(object fn, tuple args, dict kwargs, dict injections, dic
         async_to_inject = await asyncio.gather(*(provide for _, provide in to_inject_await))
         for provide, (injection, _) in zip(async_to_inject, to_inject_await):
             to_inject[injection] = provide
+    try:
+        return await fn(*args, **to_inject)
+    finally:
+        if closings:
+            for arg_key, provider in closings.items():
+                if arg_key in kwargs and isinstance(kwargs[arg_key], _Marker):
+                    continue
+                if not isinstance(provider, providers.Resource):
+                    continue
+                shutdown = provider.shutdown()
+                if _is_awaitable(shutdown):
+                    to_close_await.append(shutdown)
 
-    result = await fn(*args, **to_inject)
-
-    if closings:
-        for arg_key, provider in closings.items():
-            if arg_key in kwargs and isinstance(kwargs[arg_key], _Marker):
-                continue
-            if not isinstance(provider, providers.Resource):
-                continue
-            shutdown = provider.shutdown()
-            if _isawaitable(shutdown):
-                to_close_await.append(shutdown)
-
-        await asyncio.gather(*to_close_await)
+            await asyncio.gather(*to_close_await)
 
-    return result
 
 
-cdef bint _isawaitable(object instance):
+cdef bint _is_awaitable(object instance):
```

Comment:
This change looks unrelated to the main topic of PR. Please revert to avoid diff noise.